### PR TITLE
Issue 34 beliefs data frame init method from data frame changes the input object

### PR DESCRIPTION
**Pull request by Flix6x**
I found the same bug with initialization from Series.

### CODE REVIEW

**Comment by Flix6x** [comment]
Closes #34

**Review by nhoening — APPROVED** [review]
Good find ― an important fix.

**Comment by nhoening** [comment]
I tried to investigate why some tests are failing in this PR, but I couldn't figure it out. Maybe it is more obvious for you.

**Comment by Flix6x** [comment]
> I tried to investigate why some tests are failing in this PR, but I couldn't figure it out. Maybe it is more obvious for you.

## Observation

These tests were failing on an expect column name "reference_value" in tests where accuracy metrics were computed.

## Analysis

The BeliefsDataFrame initialization method from a Series was being triggered, while in fact initializing from a BeliefsSeries. The init method from a Series involves renaming to "event_value" in case the Series doesn't already have a name. The BeliefsSeries actually already had the name "reference_value", but that inexplicably got lost when copying (so then it was renamed and the expected name was not there any more).

## Solution

Moving the copy statement to after the conversion to a frame resolved the issue: at least column names are not lost upon copying. But instead I opted to handle BeliefsDataFrame initialization from a BeliefsSeries or BeliefsDataFrame separately. This leads me to two follow-up tickets:

## Follow-up

- [Refactor BeliefsDataFrame initialization by detecting the initialization method.](https://github.com/SeitaBV/timely-beliefs/issues/40)
- [BeliefsSeries name is lost when copying.](https://github.com/SeitaBV/timely-beliefs/issues/41)

File: timely_beliefs/beliefs/classes.py [hunk 1]
```diff
@@ -465,6 +465,14 @@ def __init__(  # noqa: C901
     ):  # noqa: C901 todo: refactor, e.g. by detecting initialization method
         """Initialise a multi-index DataFrame with beliefs about a unique sensor."""
 
+        # Initialized with a BeliefsSeries or BeliefsDataFrame
+        if len(args) > 0 and isinstance(args[0], (BeliefsSeries, BeliefsDataFrame)):
+            super().__init__(*args, **kwargs)
+            assign_sensor_and_event_resolution(
```
Comment (nhoening) [inline timely_beliefs/beliefs/classes.py:471]: It seems to me that we only call `pd.DataFrame.__init__` here, which won't understand an argument like `beliefs` or `event_start`? These are read in lines 476ff which are now never run for this case.

**Review by nhoening — CHANGES_REQUESTED** [review]

File: timely_beliefs/beliefs/classes.py (hunk 1 above)
Comment (Flix6x) [reply timely_beliefs/beliefs/classes.py:471]: That's correct. Those arguments should only play a role when initializing from a non-(BeliefsSeries or BeliefsDataFrame). That is, all information about events, sensors and sources is already contained in the BeliefsSeries or BeliefsDataFrame.

Comment (nhoening) [reply timely_beliefs/beliefs/classes.py:471]: Okay, but how is `args[0]` being preserved within `pd.DataFrame.__init__` Does that do a `self = args[0]` assignment?

Comment (Flix6x) [reply timely_beliefs/beliefs/classes.py:471]: The init methods of `pd.Series` and `pd.DataFrame` both try to extract a [block manager](https://uwekorn.com/2020/05/24/the-one-pandas-internal.html), but in different ways. What are you worried about? That the init method can still change the input in this case? Do you propose to add a copy statement here, too?

Comment (nhoening) [reply timely_beliefs/beliefs/classes.py:471]: The changing of input was not my worry, but is also interesting to check.

I didn't understand what __init__ is doing with args[0] but apparantly you are confident that the important columns (datetime and event_value) are preserved in the way we expect in a BeliefsDataFrame.

Comment (Flix6x) [reply timely_beliefs/beliefs/classes.py:471]: Yes, that I am confident of. I'll add a test for initializing a BeliefsDataFrame from a BeliefsSeres or BeliefsDataFrame.

**Pull request merged**